protocol/message_persistence.go

```diff
@@ -2207,3 +2181,161 @@ func (db sqlitePersistence) SetContactRequestState(id string, state common.Conta
+			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+		}
+	}
+	return msg, nil
+}
+
+type Messages []*common.Message
+type PinnedMessages []*common.PinnedMessage
+
+func (m Messages) Len() int            { return len(m) }
+func (m Messages) Swap(i, j int)       { m[i], m[j] = m[j], m[i] }
+func (m PinnedMessages) Len() int      { return len(m) }
+func (m PinnedMessages) Swap(i, j int) { m[i], m[j] = m[j], m[i] }
+
+type ByClock struct{ Messages }
+type ByClockPinned struct{ PinnedMessages }
+
+func (m ByClock) Less(i, j int) bool {
+	return m.Messages[i].Clock < m.Messages[j].Clock
+}
+
+func (m ByClockPinned) Less(i, j int) bool {
+	return m.PinnedMessages[i].Message.Clock < m.PinnedMessages[j].Message.Clock
+}
+
+func getMessagesFromScanRows(db sqlitePersistence, rows *sql.Rows) ([]*common.Message, error) {
+	messageIdx := make(map[string]*common.Message, 0)
+	for rows.Next() {
+		// There's a possibility of multiple rows per message if the
+		// message has a discordMessage and the discordMessage has multiple
+		// attachments
+		//
+		// Hence, we make sure we're aggregating all attachments on a single
+		// common.Message
+		var message common.Message
+		if err := db.tableUserMessagesScanAllFields(rows, &message); err != nil {
+			return nil, err
+		}
+
+		if msg, ok := messageIdx[message.ID]; !ok {
+			messageIdx[message.ID] = &message
+		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
+			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+			messageIdx[message.ID] = msg
+		}
+	}
+
+	var messages []*common.Message
+	for _, message := range messageIdx {
+		messages = append(messages, message)
+	}
+	sort.Sort(sort.Reverse(ByClock{messages}))
+
+	return messages, nil
+}
+
+func getMessagesAndCursorsFromScanRows(db sqlitePersistence, rows *sql.Rows) ([]*common.Message, []string, error) {
+
+	var cursors []string
+	messageIdx := make(map[string]*common.Message, 0)
+
+	for rows.Next() {
+		// There's a possibility of multiple rows per message if the
+		// message has a discordMessage and the discordMessage has multiple
+		// attachments
+		//
+		// Hence, we make sure we're aggregating all attachments on a single
+		// common.Message
+		var (
+			message common.Message
+			cursor  string
+		)
+		if err := db.tableUserMessagesScanAllFields(rows, &message, &cursor); err != nil {
+			return nil, nil, err
+		}
+
+		if msg, ok := messageIdx[message.ID]; !ok {
+			messageIdx[message.ID] = &message
+			cursors = append(cursors, cursor)
+		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
+			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+			messageIdx[message.ID] = msg
+		}
+	}
+
+	var messages []*common.Message
+	for _, message := range messageIdx {
+		messages = append(messages, message)
+	}
+	sort.Sort(sort.Reverse(ByClock{messages}))
+
+	return messages, cursors, nil
+}
```

Comment:
`getMessagesAndCursorsFromScanRows()` is almost identical to `getMessagesFromScanRows()`, however where it's used a `cursor` parameter in the query is required.

I could probably merge the two and add a `cursor: bool` flag or something.